packages/grid/src/vaadin-grid-selection-column-mixin.js

```diff
@@ -72,11 +74,17 @@ export const GridSelectionColumnMixin = (superClass) =>
       }
 
       if (selectAll && this.__hasArrayDataProvider()) {
-        this.__withFilteredItemsArray((items) => {
-          this._grid.selectedItems = items;
-        });
+        // Select all items that users may change the selection state of, keeping items that users may not change the selection state of
+        const newSelection = this.__getSelectableItems();
+        for (const item of this.__getNonModifiableSelectedItems()) {
+          if (!newSelection.includes(item)) {
+            newSelection.push(item);
+          }
+        }
```

Comment:
This only checks for item equality by reference, which is wrong, but is how the component currently determines the select all state as well. Was planning to fix that in a follow-up PR.

Reply:
Actually there shouldn't be an intersection between items that the user may select, and selected items that the user may not change the selection state of. Simplified this to just concat both arrays.

Reply:
This is obsolete now.